csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    public sealed partial class Communicator
+    {
+        private readonly Dictionary<(Endpoint, object?), List<Connection>> _outgoingConnections =
+            new(ConnectionMapKeyComparer.Instance);
+        private readonly Dictionary<(Endpoint, object?), Task<Connection>> _pendingOutgoingConnections =
+            new(ConnectionMapKeyComparer.Instance);
+        // We keep a map of the endpoints that recently resulted in a failure while establishing a connection. This is
```

Comment:
Is this description correct? I think logically this dictionary should contain endpoints:
- for which we tried and failed to establish a connection
- for which we received a transport exception (from an established connection)

And whenever we successfully establish a connection, we should remove its endpoint from this dictionary.

Reply:
The description is correct, endpoints in this list represent endpoints with recent failure to establish a connection